File: processing/stream-processor/src/main/java/com/chainrisk/stream/mev/model/PendingTx.java
package com.chainrisk.stream.mev.model;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;

import java.io.Serializable;
import java.math.BigInteger;

/**
 * Pending transaction from mempool
 */
@JsonIgnoreProperties(ignoreUnknown = true)
public class PendingTx implements Serializable {
    private static final long serialVersionUID = 1L;

    private String hash;
    private String from;
    private String to;
    private String value;
    private long gas;

    @JsonProperty("gas_price")
    private String gasPrice;

    @JsonProperty("gas_fee_cap")
    private String gasFeeCap;

    @JsonProperty("gas_tip_cap")
    private String gasTipCap;

    private long nonce;
    private String input;

    @JsonProperty("method_id")
    private String methodId;

    @JsonProperty("token_target")
    private String tokenTarget;

    private long timestamp;
    private String network;

    @JsonProperty("tx_type")
    private int txType;

    // Getters and setters
    public String getHash() { return hash; }
    public void setHash(String hash) { this.hash = hash; }

    public String getFrom() { return from; }
    public void setFrom(String from) { this.from = from; }

    public String getTo() { return to; }
    public void setTo(String to) { this.to = to; }

    public String getValue() { return value; }
    public void setValue(String value) { this.value = value; }

    public long getGas() { return gas; }
    public void setGas(long gas) { this.gas = gas; }

    public String getGasPrice() { return gasPrice; }
    public void setGasPrice(String gasPrice) { this.gasPrice = gasPrice; }

    public String getGasFeeCap() { return gasFeeCap; }
    public void setGasFeeCap(String gasFeeCap) { this.gasFeeCap = gasFeeCap; }

    public String getGasTipCap() { return gasTipCap; }
    public void setGasTipCap(String gasTipCap) { this.gasTipCap = gasTipCap; }

    public long getNonce() { return nonce; }
    public void setNonce(long nonce) { this.nonce = nonce; }

    public String getInput() { return input; }
    public void setInput(String input) { this.input = input; }

    public String getMethodId() { return methodId; }
    public void setMethodId(String methodId) { this.methodId = methodId; }

    public String getTokenTarget() { return tokenTarget; }
    public void setTokenTarget(String tokenTarget) { this.tokenTarget = tokenTarget; }

    public long getTimestamp() { return timestamp; }
    public void setTimestamp(long timestamp) { this.timestamp = timestamp; }

    public String getNetwork() { return network; }
    public void setNetwork(String network) { this.network = network; }

    public int getTxType() { return txType; }
    public void setTxType(int txType) { this.txType = txType; }

    /**
     * Get effective gas price for comparison
     */
    public BigInteger getEffectiveGasPrice() {
        if (txType == 2 && gasFeeCap != null && !gasFeeCap.isEmpty()) {
            return new BigInteger(gasFeeCap);
        }
        return gasPrice != null ? new BigInteger(gasPrice) : BigInteger.ZERO;
    }

    /**
     * Check if this is a DEX swap transaction
     */
    public boolean isDexSwap() {
        if (methodId == null || methodId.length() < 10) {
            return false;
        }
        String method = methodId.substring(0, 10).toLowerCase();
        return method.equals("0x38ed1739") || // swapExactTokensForTokens
               method.equals("0x8803dbee") || // swapTokensForExactTokens
               method.equals("0x7ff36ab5") || // swapExactETHForTokens
               method.equals("0x4a25d94a") || // swapTokensForExactETH
               method.equals("0x18cbafe5") || // swapExactTokensForETH
               method.equals("0xfb3bdb41") || // swapETHForExactTokens
               method.equals("0x5ae401dc") || // multicall
               method.equals("0x414bf389") || // exactInputSingle (V3)
               method.equals("0xc04b8d59") || // exactInput
               method.equals("0xdb3e2198");   // exactOutputSingle
    }

    /**
     * Extract target token from swap input data (simplified)
     */
    public String extractTargetToken() {
        if (tokenTarget != null && !tokenTarget.isEmpty()) {
            return tokenTarget;
        }
        return to; // Use contract address as fallback
    }
}

File: data-ingestion/mempool-collector/internal/model/pending_tx.go
package model

import (
	"encoding/json"
	"math/big"
)

// PendingTx represents a pending transaction from mempool
type PendingTx struct {
	Hash        string `json:"hash"`
	From        string `json:"from"`
	To          string `json:"to"`
	Value       string `json:"value"`
	Gas         uint64 `json:"gas"`
	GasPrice    string `json:"gas_price"`
	GasFeeCap   string `json:"gas_fee_cap,omitempty"`   // EIP-1559
	GasTipCap   string `json:"gas_tip_cap,omitempty"`   // EIP-1559
	Nonce       uint64 `json:"nonce"`
	Input       string `json:"input"`
	Timestamp   int64  `json:"timestamp"`
	Network     string `json:"network"`
	TxType      uint8  `json:"tx_type"`                 // 0=legacy, 2=EIP-1559
	MethodID    string `json:"method_id,omitempty"`     // first 4 bytes of input
	TokenTarget string `json:"token_target,omitempty"`  // detected token contract
}

// ToJSON serializes to JSON bytes
func (p *PendingTx) ToJSON() ([]byte, error) {
	return json.Marshal(p)
}

// ParseFromJSON deserializes from JSON bytes
func ParseFromJSON(data []byte) (*PendingTx, error) {
	var tx PendingTx
	if err := json.Unmarshal(data, &tx); err != nil {
		return nil, err
	}
	return &tx, nil
}

// GasPriceWei returns gas price as big.Int
func (p *PendingTx) GasPriceWei() *big.Int {
	val, _ := new(big.Int).SetString(p.GasPrice, 10)
	return val
}

// EffectiveGasPrice returns effective gas price for comparison
func (p *PendingTx) EffectiveGasPrice() *big.Int {
	if p.TxType == 2 && p.GasFeeCap != "" {
		val, _ := new(big.Int).SetString(p.GasFeeCap, 10)
		return val
	}
	return p.GasPriceWei()
}

// Common DEX method IDs
const (
	MethodSwapExactTokensForTokens    = "0x38ed1739"
	MethodSwapTokensForExactTokens    = "0x8803dbee"
	MethodSwapExactETHForTokens       = "0x7ff36ab5"
	MethodSwapTokensForExactETH       = "0x4a25d94a"
	MethodSwapExactTokensForETH       = "0x18cbafe5"
	MethodSwapETHForExactTokens       = "0xfb3bdb41"
	MethodMulticall                   = "0x5ae401dc"
	MethodExactInputSingle            = "0x414bf389" // Uniswap V3
	MethodExactInput                  = "0xc04b8d59"
	MethodExactOutputSingle           = "0xdb3e2198"
)

// IsDEXSwap checks if transaction is a DEX swap
func (p *PendingTx) IsDEXSwap() bool {
	if len(p.MethodID) < 10 {
		return false
	}
	switch p.MethodID[:10] {
	case MethodSwapExactTokensForTokens,
		MethodSwapTokensForExactTokens,
		MethodSwapExactETHForTokens,
		MethodSwapTokensForExactETH,
		MethodSwapExactTokensForETH,
		MethodSwapETHForExactTokens,
		MethodMulticall,
		MethodExactInputSingle,
		MethodExactInput,
		MethodExactOutputSingle:
		return true
	}
	return false
}
